File: Java_Code/Frontend/src/app/features/pending-auth/pending-auth-detail.ts
import { Component, inject, signal } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { ActivatedRoute, Router, RouterLink } from '@angular/router';

import { ApiService, errorField, errorMessage } from '../../core/api.service';
import { PendingAuthDetailView } from '../../core/models';
import { AmountPipe } from '../../shared/amount.pipe';
import { DialogComponent } from '../../shared/dialog';
import { MessageLineComponent } from '../../shared/message-line';
import { ScreenHeaderComponent } from '../../shared/screen-header';

/**
 * Authorization detail. COBOL program {@code COPAUS1C}, map {@code COPAU01 / COPAU1A}, transaction
 * {@code CPVD}.
 *
 * <p>The source legend read {@code F3=Back  F5=Mark/Remove Fraud  F8=Next Auth}. The action bar at
 * the foot of the panel carries those three and one more: the record chain runs both ways, so a
 * step backwards is offered even though the map had no key for it.</p>
 *
 * <p>Two corrections to the map. It labelled a field {@code Auth Code:} and moved the processing
 * code into it, never showing the authorization id code at all; both are shown here under their
 * own labels. And it showed the approved amount alone, so the requested amount is shown beside
 * it.</p>
 *
 * <p>F5 toggled the fraud flag on the key press alone, which on a misread screen reported the
 * wrong authorization. The intended state is confirmed here before it is sent, and an optional
 * note travels with it into the audit trail.</p>
 */
@Component({
  selector: 'cd-pending-auth-detail',
  imports: [
    FormsModule,
    RouterLink,
    AmountPipe,
    MessageLineComponent,
    ScreenHeaderComponent,
    DialogComponent,
  ],
  template: `
    <cd-screen-header
      title="Authorization details"
      subtitle="One pending authorization, with the fraud report made an explicit action"
    />

    <div class="cd-panel">
      <div class="cd-panel__head">
        <h2>Search</h2>
        <form class="cd-actions" (ngSubmit)="fetch()">
          <input
            name="accountId"
            class="cd-mono"
            style="width: 170px"
            maxlength="11"
            placeholder="Account ID"
            [class.cd-invalid]="field() === 'accountId'"
            [(ngModel)]="accountId"
            aria-label="Account number"
          />
          <input
            name="authKey"
            class="cd-mono"
            style="width: 200px"
            maxlength="14"
            placeholder="Authorization key"
            [class.cd-invalid]="field() === 'authKey'"
            [(ngModel)]="authKey"
            aria-label="Authorization key"
          />
          <button type="submit" class="cd-primary">Fetch</button>
          <button type="button" (click)="clear()">Clear</button>
          <a class="cd-btn" routerLink="/pending-authorizations">Authorization list</a>
        </form>
      </div>
    </div>

    <cd-message [text]="message()" [kind]="kind()" />

    @if (detail(); as auth) {
      <div class="cd-panel">
        <div class="cd-panel__head">
          <h2>Authorization {{ auth.transactionId }}</h2>
          <span class="cd-inline-note">Card {{ auth.cardNumber }}</span>
        </div>

        <div class="cd-panel__body">
          <div class="cd-section-title">Card and timing</div>
          <div class="cd-grid">
            <div class="cd-field">
              <label>Card number</label>
              <div class="cd-value">
                <a [routerLink]="['/cards/view', auth.cardNumber]">{{ auth.cardNumber }}</a>
              </div>
            </div>
            <div class="cd-field">
              <label>Account</label>
              <div class="cd-value">
                <a [routerLink]="['/accounts/view', auth.accountId]">{{ auth.accountId }}</a>
              </div>
            </div>
            <div class="cd-field">
              <label>Auth date</label>
              <div class="cd-value">{{ auth.authDate }}</div>
            </div>
            <div class="cd-field">
              <label>Auth time</label>
              <div class="cd-value">{{ auth.authTime }}</div>
            </div>
            <div class="cd-field">
              <label>Original date</label>
              <div class="cd-value">{{ auth.authOrigDate }}</div>
            </div>
            <div class="cd-field">
              <label>Original time</label>
              <div class="cd-value">{{ auth.authOrigTime }}</div>
            </div>
          </div>

          <div class="cd-section-title" style="margin-top: 22px">Decision</div>
          <div class="cd-grid">
            <div class="cd-field">
              <label>Auth response</label>
              <div class="cd-value">
                <span
                  class="cd-badge"
                  [class.cd-badge--ok]="auth.authRespCode === '00'"
                  [class.cd-badge--err]="auth.authRespCode !== '00'"
                >
                  {{ auth.authRespText }}
                </span>
              </div>
              <span class="cd-field__hint">Code {{ auth.authRespCode }}</span>
            </div>
            <div class="cd-field">
              <label>Response reason</label>
              <div class="cd-value">{{ auth.authRespReason }} - {{ auth.authRespReasonText }}</div>
            </div>
            <div class="cd-field">
              <label>Auth code</label>
              <div class="cd-value">{{ auth.authIdCode }}</div>
            </div>
            <div class="cd-field">
              <label>Processing code</label>
              <div class="cd-value">{{ auth.processingCode }}</div>
              <span class="cd-field__hint">
                The source screen carried this value under its Auth Code label.
              </span>
            </div>
          </div>

          <div class="cd-section-title" style="margin-top: 22px">Amount and channel</div>
          <div class="cd-grid">
            <div class="cd-field">
              <label>Approved amount</label>
              <div class="cd-value cd-value--num" [class.cd-amount-neg]="auth.approvedAmt < 0">
                {{ auth.approvedAmt | cdAmount }}
              </div>
            </div>
            <div class="cd-field">
              <label>Requested amount</label>
              <div class="cd-value cd-value--num" [class.cd-amount-neg]="auth.transactionAmt < 0">
                {{ auth.transactionAmt | cdAmount }}
              </div>
              <span class="cd-field__hint">
                The source screen showed the approved amount alone.
              </span>
            </div>
            <div class="cd-field">
              <label>POS entry mode</label>
              <div class="cd-value">
                {{ auth.posEntryMode }} &nbsp; {{ auth.posEntryModeText }}
              </div>
            </div>
            <div class="cd-field">
              <label>Source</label>
              <div class="cd-value">{{ auth.messageSource }}</div>
            </div>
            <div class="cd-field">
              <label>Message type</label>
              <div class="cd-value">{{ auth.messageType }}</div>
            </div>
          </div>

          <div class="cd-section-title" style="margin-top: 22px">Codes</div>
          <div class="cd-grid">
            <div class="cd-field">
              <label>MCC code</label>
              <div class="cd-value">{{ auth.mccCode }}</div>
            </div>
            <div class="cd-field">
              <label>Card expiry date</label>
              <div class="cd-value">{{ auth.cardExpiryDate }}</div>
            </div>
            <div class="cd-field">
              <label>Auth type</label>
              <div class="cd-value">{{ auth.authType }}</div>
            </div>
            <div class="cd-field">
              <label>Acquirer country</label>
              <div class="cd-value">{{ auth.acqrCountryCode }}</div>
            </div>
          </div>

          <div class="cd-section-title" style="margin-top: 22px">Status</div>
          <div class="cd-grid">
            <div class="cd-field">
              <label>Transaction ID</label>
              <div class="cd-value">{{ auth.transactionId }}</div>
            </div>
            <div class="cd-field">
              <label>Match status</label>
              <div class="cd-value">{{ auth.matchStatusText }}</div>
              <span class="cd-field__hint">Code {{ auth.matchStatus }}</span>
            </div>
            <div class="cd-field">
              <label>Fraud status</label>
              <div class="cd-value">
                @if (auth.authFraud === 'F') {
                  <span class="cd-badge cd-badge--err">{{ auth.fraudStatusText }}</span>
                  @if (auth.fraudRptDate) {
                    <span class="cd-inline-note">&nbsp; reported {{ auth.fraudRptDate }}</span>
                  }
                } @else {
                  {{ auth.fraudStatusText }}
                }
              </div>
            </div>
          </div>

          <div class="cd-section-title" style="margin-top: 22px">Merchant details</div>
          <div class="cd-grid cd-grid--2">
            <div class="cd-field">
              <label>Name</label>
              <div class="cd-value">{{ auth.merchantName }}</div>
            </div>
            <div class="cd-field">
              <label>Merchant ID</label>
              <div class="cd-value">{{ auth.merchantId }}</div>
            </div>
            <div class="cd-field">
              <label>City</label>
              <div class="cd-value">{{ auth.merchantCity }}</div>
            </div>
            <div class="cd-field">
              <label>State</label>
              <div class="cd-value">{{ auth.merchantState }}</div>
            </div>
            <div class="cd-field">
              <label>ZIP</label>
              <div class="cd-value">{{ auth.merchantZip }}</div>
            </div>
          </div>
        </div>

        <div class="cd-pfkeys">
          <button type="button" class="cd-danger" [disabled]="busy()" (click)="askFraud()">
            {{ auth.authFraud === 'F' ? 'Remove fraud report' : 'Mark as fraud' }}
          </button>
          <button type="button" [disabled]="!auth.previousAuthKey" (click)="previous()">
            Previous authorization
          </button>
          <button type="button" [disabled]="!auth.nextAuthKey" (click)="next()">
            Next authorization
          </button>
          <a
            class="cd-btn"
            [routerLink]="['/pending-authorizations']"
            [queryParams]="{ accountId: auth.accountId }"
          >
            Return to list
          </a>
        </div>
      </div>

      @if (confirming(); as intent) {
        <cd-dialog
          [heading]="intent === 'mark' ? 'Confirm fraud report' : 'Remove fraud report'"
          (closed)="cancelFraud()"
        >
          <p>
            @if (intent === 'mark') {
              Report authorization <strong class="cd-mono">{{ auth.authKey }}</strong> on card
              <strong class="cd-mono">{{ auth.cardNumber }}</strong> as fraud?
            } @else {
              Clear the fraud report from authorization
              <strong class="cd-mono">{{ auth.authKey }}</strong> on card
              <strong class="cd-mono">{{ auth.cardNumber }}</strong>?
            }
          </p>
          <p class="cd-inline-note">
            The source flipped this flag on the key press alone. The state you intend is sent
            explicitly, and the note below is kept with the audit record.
          </p>
          <div class="cd-field" style="margin-top: 14px">
            <label for="note">Note for the audit trail (optional)</label>
            <textarea id="note" name="note" rows="3" maxlength="200" [(ngModel)]="note"></textarea>
          </div>
          <div class="cd-actions" style="margin-top: 14px">
            <button type="button" class="cd-danger" [disabled]="busy()" (click)="applyFraud()">
              {{ intent === 'mark' ? 'Yes, mark as fraud' : 'Yes, remove the report' }}
            </button>
            <button type="button" [disabled]="busy()" (click)="cancelFraud()">Cancel</button>
          </div>
        </cd-dialog>
      }
    }
  `,
})
export class PendingAuthDetailComponent {
  private readonly api = inject(ApiService);
  private readonly route = inject(ActivatedRoute);
  private readonly router = inject(Router);

  accountId = '';
  authKey = '';
  note = '';

  readonly detail = signal<PendingAuthDetailView | null>(null);
  readonly message = signal<string | null>(null);
  readonly kind = signal<'error' | 'info' | 'ok' | 'warn'>('error');
  readonly field = signal<string | null>(null);

  /** Which way the fraud flag is about to be moved, and nothing while no answer is being asked. */
  readonly confirming = signal<'mark' | 'remove' | null>(null);
  readonly busy = signal(false);

  constructor() {
    const key = this.route.snapshot.paramMap.get('authKey');
    const account = this.route.snapshot.queryParamMap.get('accountId');
    if (key) {
      this.authKey = key;
    }
    if (account) {
      this.accountId = account;
    }
    if (key && account) {
      this.load(account, key);
    }
  }

  /** ENTER: an account and an authorization key together name one record. */
  fetch(): void {
    this.message.set(null);
    this.field.set(null);
    this.confirming.set(null);
    const account = this.accountId.trim();
    const key = this.authKey.trim();
    if (!account) {
      this.reject('Please enter Acct Id...', 'accountId');
      return;
    }
    if (!key) {
      this.reject('Please select an authorization...', 'authKey');
      return;
    }
    this.load(account, key);
  }

  /** Empties the screen. F3 returns to the authorization list, which is the link beside this. */
  clear(): void {
    this.accountId = '';
    this.authKey = '';
    this.note = '';
    this.detail.set(null);
    this.confirming.set(null);
    this.message.set(null);
    this.field.set(null);
  }

  /** F8 advances to the next authorization under the same account. */
  next(): void {
    const auth = this.detail();
    if (!auth) {
      return;
    }
    if (!auth.nextAuthKey) {
      // The button is already disabled at the end of the chain. The source literal still answers
      // any other route to this action rather than the press appearing to do nothing.
      this.kind.set('info');
      this.message.set('Already at the last Authorization...');
      return;
    }
    this.step(auth.accountId, auth.nextAuthKey);
  }

  /** Steps back one authorization. The source had no backward key; the record chain has one. */
  previous(): void {
    const auth = this.detail();
    if (!auth) {
      return;
    }
    if (!auth.previousAuthKey) {
      this.kind.set('info');
      this.message.set('Already at the first Authorization...');
      return;
    }
    this.step(auth.accountId, auth.previousAuthKey);
  }

  /**
   * F5 marks or removes the fraud report. The source toggled the flag as the key was pressed, so
   * the intent is put to the user first and the request carries it rather than a toggle.
   */
  askFraud(): void {
    const auth = this.detail();
    if (!auth) {
      return;
    }
    this.message.set(null);
    this.field.set(null);
    this.note = '';
    this.confirming.set(auth.authFraud === 'F' ? 'remove' : 'mark');
  }

  /** Sends the confirmed state: marked when the report is being made, cleared when it is lifted. */
  applyFraud(): void {
    const auth = this.detail();
    const intent = this.confirming();
    if (!auth || !intent) {
      return;
    }
    const note = this.note.trim();
    this.busy.set(true);
    this.message.set(null);
    this.field.set(null);
    this.api
      .markAuthorizationFraud(auth.accountId, auth.authKey, {
        confirmed: intent === 'mark',
        note: note || undefined,
      })
      .subscribe({
        next: (result) => {
          this.busy.set(false);
          this.confirming.set(null);
          this.note = '';
          this.detail.set(result.detail);
          this.kind.set('ok');
          this.message.set(result.message);
        },
        error: (error: unknown) => {
          this.busy.set(false);
          this.confirming.set(null);
          this.kind.set('error');
          this.message.set(errorMessage(error));
          this.field.set(errorField(error));
        },
      });
  }

  /** Leaves the flag as it stands. */
  cancelFraud(): void {
    this.confirming.set(null);
    this.note = '';
  }

  /** Moves along the chain, clearing whatever the previous record left on the message line. */
  private step(accountId: string, authKey: string): void {
    this.message.set(null);
    this.field.set(null);
    this.confirming.set(null);
    this.load(accountId, authKey);
  }

  private load(accountId: string, authKey: string): void {
    this.api.pendingAuthorization(accountId, authKey).subscribe({
      next: (detail) => {
        this.detail.set(detail);
        this.accountId = detail.accountId;
        this.authKey = detail.authKey;
        this.syncUrl(detail);
      },
      error: (error: unknown) => {
        this.detail.set(null);
        this.kind.set('error');
        this.message.set(errorMessage(error));
        this.field.set(errorField(error));
      },
    });
  }

  /** Reports a client side edit with the source literal and highlights the offending box. */
  private reject(text: string, field: string): void {
    this.detail.set(null);
    this.kind.set('error');
    this.message.set(text);
    this.field.set(field);
  }

  /**
   * Keeps the address bar on the record being shown, so an authorization reached by stepping along
   * the chain is still linkable. The entry is replaced rather than pushed, which leaves Back
   * pointing at the list the screen was opened from rather than at every record walked through.
   */
  private syncUrl(detail: PendingAuthDetailView): void {
    this.router.navigate(['/pending-authorizations/view', detail.authKey], {
      queryParams: { accountId: detail.accountId },
      replaceUrl: true,
    });
  }
}

File: Java_Code/Backend/src/main/java/com/carddemo/domain/PendingAuthDetail.java
package com.carddemo.domain;

import jakarta.persistence.Column;
import jakarta.persistence.Embeddable;
import jakarta.persistence.EmbeddedId;
import jakarta.persistence.Entity;
import jakarta.persistence.PrePersist;
import jakarta.persistence.PreUpdate;
import jakarta.persistence.Table;
import jakarta.persistence.Version;
import lombok.AllArgsConstructor;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.io.Serializable;
import java.math.BigDecimal;
import java.time.LocalDateTime;

/**
 * One pending authorization. COBOL source {@code CIPAUDTY.cpy} / IMS segment
 * {@code PAUTDTL1} (child of {@code PAUTSUM0}, 200 bytes) of database {@code DBPAUTP0}.
 *
 * <p>Byte offsets within the segment: authorization key 1-8, original date 9-14, original
 * time 15-20, card number 21-36, authorization type 37-40, card expiry 41-44, message type
 * 45-50, message source 51-56, authorization id code 57-62, response code 63-64, response
 * reason 65-68, processing code 69-74, transaction amount 75-81 ({@code S9(10)V99 COMP-3}),
 * approved amount 82-88, merchant category code 89-92, acquirer country 93-95, POS entry
 * mode 96-97, merchant id 98-112, merchant name 113-134, merchant city 135-147, merchant
 * state 148-149, merchant ZIP 150-158, transaction id 159-173, match status 174, fraud flag
 * 175, fraud report date 176-183, filler 184-200.</p>
 *
 * <p><b>The key is a nine's complement.</b> {@code PA-AUTHORIZATION-KEY} is
 * {@code PA-AUTH-DATE-9C S9(5) COMP-3} followed by {@code PA-AUTH-TIME-9C S9(9) COMP-3},
 * and the {@code 9C} suffix is what it says: the stored digits are 99999 minus the Julian
 * date and 999999999 minus the time, so an ascending DL/I scan returns the newest
 * authorization first. {@link Key#authKey} keeps those complement digits verbatim as a
 * fixed width 14 character string, which makes {@code ORDER BY auth_key} reproduce the
 * physical IMS order that {@code COPAUS0C} and {@code COPAUS1C} walk with {@code GNP}.</p>
 */
@Entity
@Table(name = "pending_auth_detail")
@Getter
@Setter
@NoArgsConstructor
public class PendingAuthDetail {

    /** Values of {@code PA-MATCH-STATUS}. */
    public static final String MATCH_PENDING = "P";
    public static final String MATCH_AUTH_DECLINED = "D";
    public static final String MATCH_PENDING_EXPIRED = "E";
    public static final String MATCHED_WITH_TRAN = "M";

    /** Values of {@code PA-AUTH-FRAUD}. */
    public static final String FRAUD_CONFIRMED = "F";
    public static final String FRAUD_REMOVED = "R";

    @EmbeddedId
    private Key id;

    /** The Julian date the complement key decodes to, {@code yyddd}. */
    @Column(name = "auth_julian_date", nullable = false)
    private int authJulianDate;

    /** The time the complement key decodes to, as its nine digit numeric value. */
    @Column(name = "auth_time_value", nullable = false)
    private int authTimeValue;

    /** {@code PA-AUTH-ORIG-DATE} X(6), {@code yymmdd}. */
    @Column(name = "auth_orig_date", length = 6, nullable = false)
    private String authOrigDate = "";

    /** {@code PA-AUTH-ORIG-TIME} X(6), {@code hhmmss}. */
    @Column(name = "auth_orig_time", length = 6, nullable = false)
    private String authOrigTime = "";

    @Column(name = "card_number", length = 16, nullable = false)
    private String cardNumber = "";

    @Column(name = "auth_type", length = 4, nullable = false)
    private String authType = "";

    @Column(name = "card_expiry_date", length = 4, nullable = false)
    private String cardExpiryDate = "";

    @Column(name = "message_type", length = 6, nullable = false)
    private String messageType = "";

    @Column(name = "message_source", length = 6, nullable = false)
    private String messageSource = "";

    @Column(name = "auth_id_code", length = 6, nullable = false)
    private String authIdCode = "";

    /** {@code PA-AUTH-RESP-CODE} X(2); {@code 00} is the approved condition name. */
    @Column(name = "auth_resp_code", length = 2, nullable = false)
    private String authRespCode = "";

    @Column(name = "auth_resp_reason", length = 4, nullable = false)
    private String authRespReason = "";

    @Column(name = "processing_code", length = 6, nullable = false)
    private String processingCode = "";

    /** {@code PA-TRANSACTION-AMT} S9(10)V99 COMP-3. */
    @Column(name = "transaction_amt", nullable = false, precision = 14, scale = 2)
    private BigDecimal transactionAmt = BigDecimal.ZERO;

    /** {@code PA-APPROVED-AMT} S9(10)V99 COMP-3. */
    @Column(name = "approved_amt", nullable = false, precision = 14, scale = 2)
    private BigDecimal approvedAmt = BigDecimal.ZERO;

    /** {@code PA-MERCHANT-CATAGORY-CODE} X(4), the source spelling. */
    @Column(name = "mcc_code", length = 4, nullable = false)
    private String mccCode = "";

    @Column(name = "acqr_country_code", length = 3, nullable = false)
    private String acqrCountryCode = "";

    /** {@code PA-POS-ENTRY-MODE} 9(2), kept as text so a leading zero survives. */
    @Column(name = "pos_entry_mode", length = 2, nullable = false)
    private String posEntryMode = "";

    @Column(name = "merchant_id", length = 15, nullable = false)
    private String merchantId = "";

    @Column(name = "merchant_name", length = 22, nullable = false)
    private String merchantName = "";

    @Column(name = "merchant_city", length = 13, nullable = false)
    private String merchantCity = "";

    @Column(name = "merchant_state", length = 2, nullable = false)
    private String merchantState = "";

    @Column(name = "merchant_zip", length = 9, nullable = false)
    private String merchantZip = "";

    @Column(name = "transaction_id", length = 15, nullable = false)
    private String transactionId = "";

    /** {@code PA-MATCH-STATUS} X(1): P, D, E or M. */
    @Column(name = "match_status", length = 1, nullable = false)
    private String matchStatus = "";

    /** {@code PA-AUTH-FRAUD} X(1): F confirmed, R removed, blank never reported. */
    @Column(name = "auth_fraud", length = 1, nullable = false)
    private String authFraud = "";

    /** {@code PA-FRAUD-RPT-DATE} X(8). */
    @Column(name = "fraud_rpt_date", length = 8, nullable = false)
    private String fraudRptDate = "";

    @Column(name = "created_at", nullable = false)
    private LocalDateTime createdAt;

    @Column(name = "updated_at", nullable = false)
    private LocalDateTime updatedAt;

    @Version
    @Column(name = "version", nullable = false)
    private long version;

    public PendingAuthDetail(String accountId, String authKey) {
        this.id = new Key(accountId, authKey);
    }

    public String getAccountId() {
        return id == null ? null : id.getAccountId();
    }

    public String getAuthKey() {
        return id == null ? null : id.getAuthKey();
    }

    /** True when this authorization is currently reported as fraud. */
    public boolean isFraudConfirmed() {
        return FRAUD_CONFIRMED.equals(authFraud);
    }

    @PrePersist
    void onCreate() {
        createdAt = LocalDateTime.now();
        updatedAt = createdAt;
    }

    @PreUpdate
    void onUpdate() {
        updatedAt = LocalDateTime.now();
    }

    @Embeddable
    @Getter
    @Setter
    @NoArgsConstructor
    @AllArgsConstructor
    @EqualsAndHashCode
    public static class Key implements Serializable {

        /** The parent {@code PAUTSUM0} account id. */
        @Column(name = "account_id", length = 11, nullable = false)
        private String accountId;

        /** {@code PAUT9CTS}: five complement date digits then nine complement time digits. */
        @Column(name = "auth_key", length = 14, nullable = false)
        private String authKey;
    }
}
